darts/utils/data/sequential_dataset.py

```diff
@@ -33,86 +100,262 @@ def __init__(self,
 
-        # The time series index of our forecasting point (indexed from the end of the series):
-        forecast_point_idx = self.output_chunk_length + lh_idx
+        The sampling is uniform over the number of time series; i.e., the i-th sample of this dataset has
+        a probability 1/N of coming from any of the N time series in the sequence. If the time series have different
+        lengths, they will contain different numbers of slices. Therefore, some particular slices may
+        be sampled more often than others if they belong to shorter time series.
 
-        # select input and outputs, using the previously computed indexes
-        input_target = ts_target[-(forecast_point_idx + self.input_chunk_length):-forecast_point_idx]
-        if forecast_point_idx == self.output_chunk_length:
-            # we need this case because "-0" is not supported as an indexing bound
-            output_target = ts_target[-forecast_point_idx:]
-        else:
-            output_target = ts_target[-forecast_point_idx:-forecast_point_idx + self.output_chunk_length]
+        Parameters
+        ----------
+        target_series
+            One or a sequence of target `TimeSeries`.
+        past_covariates
+            Optionally, one or a sequence of `TimeSeries` containing past-observed covariates. If this parameter is set,
+            the provided sequence must have the same length as that of `target_series`. Moreover, all
+            covariates in the sequence must have a time span large enough to contain all the required slices.
+            The joint slicing of the target and covariates is relying on the time axes of both series.
+        future_covariates
+            Optionally, one or a sequence of `TimeSeries` containing future-known covariates. This has to follow
+            the same constraints as `past_covariates`.
+        input_chunk_length
+            The length of the emitted past series.
+        output_chunk_length
+            The length of the emitted future series.
+        max_samples_per_ts
+            This is an upper bound on the number of tuples that can be produced per time series.
+            It can be used in order to have an upper bound on the total size of the dataset and
+            ensure proper sampling. If `None`, it will read all of the individual time series in advance (at dataset
+            creation) to know their sizes, which might be expensive on big datasets.
+            If some series turn out to have a length that would allow more than `max_samples_per_ts`, only the
+            most recent `max_samples_per_ts` samples will be considered.
+        """
+        super().__init__()
 
-        # optionally also produce the input covariate
-        input_covariate = None
-        if self.covariates is not None:
-            ts_covariate = self.covariates[ts_idx].values(copy=False)
+        self.ds_past = GenericShiftedDataset(target_series=target_series,
```

Comment:
I like this organization of breaking a bigger dataset into 2 smaller ones, seems like a clean approach! But I guess the name `ds_past` isn't pefectly accurate since it includes 'future' information in the form of the future target. Also, it might be a bit arbitrary that the past covariates is included in the dataset with the target, and all other types of covariates require their specialized sub-dataset. But of course I assume that is a product of our existing codebase, and we can reorganize this in later PRs too!

Reply:
@pennfranc I'm not sure to exactly understand what you mean. `ds_past` is named like that simply because it will be the "internal dataset" in charge of serving the past covariates. The past and future target is served both by `ds_past` and `ds_future`, so we make the choice to use the one from `ds_past` and drop those of `ds_future` (naming them `_`). The sole purpose of `ds_future` is to serve `future_covariates`. Behind the scenes the target will be sliced one time too many, but I think it's (hopefully) negligible overhead compared to the code simplicity it brings.
But perhaps I'm missing something in your comment?

Reply:
Yeah I think your logic makes sense. I guess the first thought I had when seeing the names was that `ds_past` would supply all past data (both target and covariate), and `ds_future` would provide all future data. What do you think about adding this explanation of the two datasets in the form of comments?